### PR TITLE
Master is now Py >=3.6 ONLY

### PR DESCRIPTION
**Pull request by nabobalis**
Parts of this PR:

1. Moved to upstream/master version of astropy_helpers
2. Moved Figure tests to Py3.6
3. Removed old metaclasses for Map and Time Series now we can use Py3.6 only magic. 
4. General expansion/update of the gitignore
5. NOT ANYMORE
6. Probs other things I forget now.

### CODE REVIEW

File: sunpy/map/mapbase.py [hunk 1]
```diff
@@ -175,6 +136,13 @@ class GenericMap(NDData):
         rest will be discarded.
     """
 
+    _registry = dict()
+
+    def __init_subclass__(cls, **kwargs):
```
Comment (Cadair) [inline sunpy/map/mapbase.py]: Given this is pretty arcane and also a very new Python feature a good docstring describing what this is doing in general and for us is probably needed.

File: sunpy/timeseries/timeseriesbase.py [hunk 2]
```diff
@@ -111,6 +73,13 @@ class GenericTimeSeries:
 
     # Class attribute used to specify the source class of the TimeSeries.
     _source = None
+    _registry = dict()
+
+    def __init_subclass__(cls, **kwargs):
```
Comment (Cadair) [inline sunpy/timeseries/timeseriesbase.py]: same as above

File: sunpy/timeseries/timeseries_factory.py [hunk 3]
```diff
@@ -565,6 +565,5 @@ class NoTimeSeriesFound(ValueError):
     pass
 
 
-TimeSeries = TimeSeriesFactory(default_widget_type=GenericTimeSeries,
+TimeSeries = TimeSeriesFactory(registry= GenericTimeSeries._registry, default_widget_type=GenericTimeSeries,
```
Comment (Cadair) [inline sunpy/timeseries/timeseries_factory.py]: rouge space

File: sunpy/map/mapbase.py (hunk 1 above)
Comment (nabobalis) [reply sunpy/map/mapbase.py]: Fixed. ON THE LINE BELOW.

File: sunpy/timeseries/timeseriesbase.py (hunk 2 above)
Comment (nabobalis) [reply sunpy/timeseries/timeseriesbase.py]: same as above

**Comment by Cadair** [comment]
@pep8speaks quiet

**Comment by Cadair** [comment]
@ayshih underneath the other changes in here is some changes to the metaclass stuff that you might be interested in reviewing.

File: ah_bootstrap.py [hunk 4]
```diff
@@ -38,63 +38,42 @@
...
 import errno
-import imp
 import io
 import locale
 import os
 import re
 import subprocess as sp
 import sys
 
+__minimum_python_version__ = (3, 5)
```
Comment (dstansby) [inline ah_bootstrap.py]: Should this be `(3, 6)`? (is this file just an update from an external source?)

File: changelog/2655.trivial.rst [hunk 5]
```diff
@@ -0,0 +1,4 @@
+Removed old metaclass used for Map and TimeSeries as we have now moved to Python 3.6.
+Moved figure tests to 3.6.
```
Comment (dstansby) [inline changelog/2655.trivial.rst]: "3.6" > "Python 3.6"

File: sunpy/tests/figure_tests_env_py36.yml [hunk 6]
```diff
@@ -0,0 +1,26 @@
+name: sunpy-figure-tests-3.6
```
Comment (dstansby) [inline sunpy/tests/figure_tests_env_py36.yml]: Is it possible to keep the requirements the same as before, but just bump the python version? Then we can start bumping e.g. Matplotlib requirements when I have image comparison working and we can see what the changes are.

File: ah_bootstrap.py (hunk 4 above)
Comment (Cadair) [reply ah_bootstrap.py]: yeah this is just an update from astropy_helpers, we will add our own version of this when I get around to it lol.

File: sunpy/tests/figure_tests_env_py36.yml (hunk 6 above)
Comment (nabobalis) [reply sunpy/tests/figure_tests_env_py36.yml]: I didn't think to do that. I just updated my current env, saved it to a file and then updated the versions manually. 

The image comparison I used was the figures on your repo, everything bar the issue I opened looked the same dispite the hash change.

**Comment by Cadair** [comment]
We should move to a tagged version of astropy_helpers, I.e. a release.

File: sunpy/map/mapbase.py [hunk 7]
```diff
@@ -43,41 +39,6 @@
 __all__ = ['GenericMap']
 
 
-"""
-Questions
----------
-* Should we use Helioviewer or VSO's data model? (e.g. map.meas, map.wavelength
-or something else?)
-* Should 'center' be renamed to 'offset' and crpix1 & 2 be used for 'center'?
-"""
```
Comment (nabobalis) [inline sunpy/map/mapbase.py]: What should we do about these?

File: sunpy/database/commands.py [hunk 8]
```diff
@@ -1,20 +1,9 @@
-# Author: Simon Liedtke <liedtke.simon@googlemail.com>
-#
-# This module was developed with funding provided by
-# the Google Summer of Code (2013).
-
-from __future__ import absolute_import
-
-from abc import ABCMeta, abstractmethod
+from abc import ABC, abstractmethod
```
Comment (Cadair) [inline sunpy/database/commands.py]: this should be below the `os` import

File: sunpy/map/mapbase.py (hunk 7 above)
Comment (Cadair) [reply sunpy/map/mapbase.py]: They are super old we can ditch them

File: sunpy/tests/helpers.py [hunk 9]
```diff
@@ -2,10 +2,10 @@
 # Lovingly borrowed from Astropy
 # Licensed under a 3-clause BSD style license - see licences/ASTROPY.rst
 
+import warnings
```
Comment (Cadair) [inline sunpy/tests/helpers.py]: these should be where they were before so they are in length then alphabetical order. #fussy

File: sunpy/tests/figure_tests_env_py36.yml (hunk 6 above)
Comment (nabobalis) [reply sunpy/tests/figure_tests_env_py36.yml]: I did as you suggested. The hashes are the same but new python version.

**Comment by nabobalis** [comment]
@ayshih I will take that as an approval of the PR!

**Comment by ayshih** [comment]
You can if you wish. =)  I only really looked at the metaclass stuff, but everything else seems fine at first look.

**Comment by nabobalis** [comment]
Nope, I need a second review so come on and step up @sunpy/sunpy-developers!

**Pull request merged**